Progress #1142 -- Consumables now have their stack count reduced when used

cabeca1143 opened the pull request:
* Introduced ItemClass parameter to ItemData in order to differenciate consumables from other items.
* Introduced two GetItem(string itemSpellName) in Inventory and InventoryManager in order to properly assign an item with just it's spellName string
* Introduced a check in HandleCastSpell.cs to check if the spell being casted originates from an item, and another check to see if it is a consumable in order to proceed with the stack removal.

Props to lizardy for his guidance.

Progress #1142

cabeca1143 committed:
Update

cabeca1143 committed:
removed unused variable

cabeca1143 committed:
Removed some summaries i used when making changes

lzardy commented on GameServerCore/Domain/IInventoryManager.cs:
```diff
@@ -1,14 +1,17 @@
     public interface IInventoryManager
     {
         IItem GetItem(byte slot);
+        IItem GetItem(string ItemName);
```
`ItemName` -> `ItemSpellName`

cabeca1143 replied:
resolved

lzardy commented on GameServerLib/Items/Inventory.cs:
```diff
@@ -130,7 +143,13 @@ private IItem AddStackingItem(IItemData item)
+        {
+            IItem item = GetItem(itemSpellName);
 
+            game.PacketNotifier.NotifyRemoveItem(owner, GetItemSlot(item), (byte)(item.StackCount - 1));
```
Why not Decrement before the packet?

cabeca1143 replied:
Because it would cause a crash with 1 stack, since the item would be removed, resulting on a null exception

lzardy replied:
Resolved then.

cabeca1143 replied:
So, i found out that it wasn't crashing because the item was null, is was because it was casting a BaseSpell

lzardy commented on GameServerLib/Items/InventoryManager.cs:
```diff
@@ -29,16 +32,31 @@ public IItem GetItem(byte slot)
             return _inventory.GetItem(slot);
         }
 
+        public IItem GetItem(string itemSpellName)
```
Replace body of function with `return _inventory.GetItem(itemSpellName);`

cabeca1143 replied:
resolved

lzardy replied:
Not actually resolved.

cabeca1143 replied:
Oops my bad

lzardy commented on GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs:
```diff
@@ -35,6 +35,15 @@ public override bool HandlePacket(int userId, CastSpellRequest req)
+            if (s.CastInfo.SpellSlot == (int)SpellSlotType.InventorySlots)
+            {
+                var item = s.CastInfo.Owner.Inventory.GetItem(s.SpellName);
+                if (item.ItemData.ItemClass == 1)
```
`ItemClass` variable does not reflect consumability.
Use `Consumed` variable instead.

cabeca1143 replied:
resolved

lzardy reviewed (changes requested):
Changes needed.

lzardy commented:
Fix Tests as well for AppVeyor.

danil179 commented on GameServerLib/Items/Inventory.cs:
```diff
@@ -130,7 +143,13 @@ private IItem AddStackingItem(IItemData item)
+        {
+            IItem item = GetItem(itemSpellName);
 
+            game.PacketNotifier.NotifyRemoveItem(owner, GetItemSlot(item), (byte)(item.StackCount - 1));
```
PacketNotifier and Game class shouldn't be in Inventory. ~~The packets should be sent after the inventory has been updated in each event (and there are only few: Buy, sell and use).~~

InventoryManger should handle all packet related stuff.

lzardy replied:
Something you could do is add a _toRemove boolean variable to Item.cs, then inside ObjectManager.cs, check each unit with an inventory for removed items, and call the packets there.

cabeca1143 replied:
resolved

danil179 reviewed (changes requested):
Remove Game class everywhere it was added and PacketNotifier from Inventory.

EDIT:
After rechecking the design, I think that the packets should be handled in InventoryManger for all packets related stuff. AddItem has to call BuyItem (add new spells if necessary and modifiers if exist and so on, just copy from Shop) and remove item should do the same. Please only pass PacketManager to this class to avoid propagating Game class everywhere.

cabeca1143 committed:
Requested changes + Fix

cabeca1143 committed:
Moved a bunch of shop stuff to Inventory and InventoryManager

cabeca1143 committed:
Some more changes + Fix on the test

cabeca1143 committed:
Moved packet over to InventoryManager from Inventory

lzardy commented on GameServerLib/Packets/PacketHandlers/HandleCastSpell.cs:
```diff
@@ -35,6 +35,15 @@ public override bool HandlePacket(int userId, CastSpellRequest req)
                 return false;
             }
 
+            if (s.CastInfo.SpellSlot >= (int)SpellSlotType.InventorySlots && s.CastInfo.SpellSlot < (int)SpellSlotType.BluePillSlot)
```
Change to `== InventorySlots`

cabeca1143 replied:
It is >= so it gets everything in-between InventorySlots and BluePillSlot.
Setting it to == InventorySlots will make it only be abble to detect slot 6.

lzardy replied:
Nevermind then ¯\ _(ツ)_/¯

cabeca1143 replied:
I just noticed this function is somehow being called twice for some reason

lzardy replied:
It normally does for item actives, although I dont know the cause.

cabeca1143 replied:
"Normally" xD, well at least now i know it wasn't my oopsie hahaha.
I'll investigate it a bit, but i'm just too exausted to dive into it at this point

danil179 reviewed (changes requested):
The shop functions for adding/removing items should replace exactly the functions in InventoryManager. There shouldn't be multiple functions to do related things, but instead there should be a simple mechanism to handle them all.

I don't think that `RemoveStackingItem` is really needed if RemoveItem is written correctly. The behavior is the same as the one in Shop.

Also, InventoryManager should at least hold the owner as IObjAiBase (see #527) in the class variables.

Summary:
Two functions in InventoryManager:
bool AddItem(IItem item) [Returns false if the item cannot be added (e.g. inventory is full). The class should handle the stacking automatically if it is consumable by stacking the first item.]
bool RemoveItem(byte slotId, int stacksToRemove=1) [Returns false if the slot was empty at the start. If consumable, removes only stacksToRemove stacks.]
bool RemoveItem(IItemData itemData) [Find the first item to match the data (using GetItem or so) and remove it. Then, returns false if the slot was empty at the start. Should be implemented as return RemoveItem(GetItem(itemData)) or similar]

These two functions (and the helper one) should do the same functionality as the one removed from Shop currently.

cabeca1143 committed:
Requested changes

cabeca1143 committed:
Missed change

cabeca1143 commented:
Judging by the ammount of errors on GameServerLibTest, i'd say changing AddItem to bool wasnt a good idea, i'll work on something else

cabeca1143 committed:
More requested changes

*Added RemoveItem overload with an Item parameter back.

*Turned both RemoveItems into a bool function and added an extra "stacksToRemove" parameter.

Changed the "DecreaseStackCount" function in order for it to be able to take in an Int Parameter

*Turned AddItem in a tuple function in order to return IItem and a bool

Updated the tests in order to reflect the AddItem tuple change

lzardy commented on GameServerCore/Domain/IInventoryManager.cs:
```diff
@@ -1,15 +1,18 @@
-        IItem AddItem(IItemData item);
+        bool RemoveItem(byte slot, IObjAiBase owner = null, int stacksToRemove = 1);
+        bool RemoveItem(IItem item, IObjAiBase owner = null, int stacksToRemove = 1);
+        Tuple<IItem, bool> AddItem(IItemData item, IObjAiBase owner = null);
```
Use `KeyValuePair<INTEL, bool>`

cabeca1143 replied:
Resolved

lzardy commented on GameServerCore/Domain/IStackable.cs:
```diff
@@ -5,7 +5,7 @@ public interface IStackable
         int StackCount { get; }
         bool IncrementStackCount();
-        bool DecrementStackCount();
+        bool DecrementStackCount(int stacksToRemove = 0);
```
Decrement is the opposite of Increment, which goes up by one. Instead of using Decrement, use SetStacks and revert the changes to Decrement.

cabeca1143 replied:
Resolved

lzardy commented on GameServerLib/GameObjects/Other/Stackable.cs:
```diff
@@ -19,14 +19,27 @@ public virtual bool IncrementStackCount()
         }
 
-        public virtual bool DecrementStackCount()
+        public virtual bool DecrementStackCount(int stacksToRemove = 0)
```
Revert and use SetStacks instead.

cabeca1143 replied:
Using SetStacks conflicts with the `if (newStacks < 1 || newStacks > MaxStacks)` when setting the stacks to 0 in order to remove the item, i'd have to delete that, if that's alright

lzardy replied:
Simply check in RemoveItem before calling SetStacks.

cabeca1143 replied:
Resolved

lzardy commented on GameServerLib/Items/Inventory.cs:
```diff
@@ -60,14 +75,44 @@ public IItem GetItem(byte slot)
+        public void RemoveItem(byte slot, IObjAiBase owner, int stacksToRemove)
         {
-            RemoveItem(GetItemSlot(item));
+            Items[slot].DecrementStackCount(stacksToRemove);
```
Use SetStacks instead.

lzardy commented on GameServerLib/Items/InventoryManager.cs:
```diff
@@ -1,22 +1,40 @@
         }
 
-        public IItem AddItem(IItemData item)
+        public Tuple<IItem, bool> AddItem(IItemData itemData, IObjAiBase owner = null)
```
Use `KeyValuePair<INTEL, bool>`

lzardy reviewed (changes requested):
Some changes needed.

cabeca1143 committed:
Requested Changes

*Changed AddItem from Tuple to KeyValuePair.

*Reverted DecreaseStackCount function

*RemoveItem now uses SetStacks instead of DecreaseStackCount

danil179 commented on GameServerLib/Items/Shop.cs:
```diff
@@ -56,55 +54,19 @@ public bool HandleItemBuyRequest(int itemId)
                 }
 
-                foreach (var item in ownedItems)
+                foreach (var itens in ownedItems)
```
Should be `items`

cabeca1143 replied:
> Generally approved. I'll have to test this and check if it works. I'll open an issue about passing the owner as parameter. I think that a class variable is the correct solution (and also merging InventoryManager with Inventory at some point as it is not clear why these two got separated). Is there any scenario that owner is actually needed to be `null`?

I don't remember any actual case in-game they were `null`. I added it because`TestInventoryManager.cs` doesn't contain have any IObjAiBase to be used as an owner, so i thought it would be best to go the "safe" way and make an exeception for it

danil179 reviewed (commented):
Generally approved. I'll have to test this and check if it works. I'll open an issue about passing the owner as parameter. I think that a class variable is the correct solution (and also merging InventoryManager with Inventory at some point as it is not clear why these two got separated). Is there any scenario that owner is actually needed to be `null`?

cabeca1143 committed:
Requested change + missed AddItem change on turrets

cabeca1143 committed:
Fixed turret item stats being added twice

danil179 reviewed (approved):
Approved. There are many problems with casting and so, but it is not related to the current fix. There are visual problems after consuming the last stack for some unknown reason and the possibility to consume consumables after they have depleted due to this bug. I'll open an issue about an hotfix for that, but the problem is in the casting system and client<-> server miscommunication probably.
